File: src/main/java/oscar/cp/constraints/CPObjectiveMaximize.java
/*******************************************************************************
 * This file is part of OscaR (Scala in OR).
 *   
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *  
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *  
 * You should have received a copy of the GNU General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/gpl-3.0.html
 ******************************************************************************/
package oscar.cp.constraints;


import oscar.cp.core.CPVarInt;

/**
 * @author Pierre Schaus pschaus@gmail.com
 */
public class CPObjectiveMaximize extends CPObjective{
	
	private CPVarInt objVar;
	
	public CPObjectiveMaximize(CPVarInt objVar) {
		super(objVar);
		this.objVar = objVar;
		optimum = objVar.getMax();
	}

	@Override
	public void tighten() throws RuntimeException{
		if (!objVar.isBound()) {
			throw new RuntimeException("objective not bound, not possible to tighten");
		}
		setNewBound(Math.max(objVal+1, objVar.getValue()+1));
		System.out.println("objective tighten to "+(objVal-1));
	}
	
	@Override
	public void relax() {
		objVal = Integer.MIN_VALUE;
	}

	@Override
    public boolean isOptimum() {
        return getBound() > getOptimumBound();
    }

	@Override
	protected CPObjectiveConstraint createObjectiveConstraint(CPVarInt objVar) {
		return new CPObjectiveConstraintMaximize(this, objVar);
	}



}

File: src/main/scala/oscar/cp/constraints/CPObjectiveMaximize.scala
package oscar.cp.constraints
import oscar.cp.core._
import oscar.cp.core.CPOutcome
import oscar.cp.core.CPPropagStrength

class CPObjectiveMaximize(objVars: CPVarInt*) extends CPObjective(objVars:_*) {
  
	for (i <- 0 until bounds.size) {
      bounds(i) = Int.MinValue
      bestObjs(i) = Int.MinValue
    }
	
	private val upperBounds = Array.tabulate(objVars.size)(i => objVars(i).min)
  
  	override def tighten() = {
  	  if (!currentObjective.isBound) {
  	    throw new RuntimeException("objective not bound:" + currentObjective)
  	  }
  	  if (!s.silent) println("objective "+currentObjectiveIdx+" tightened to "+(bound).max(currentObjective.value)+" ub:"+upperBounds(currentObjectiveIdx))
  	  for (i <- 0 until objVars.size) {
  	    bestObjs(i) = objVars(i).value
  	    bounds(i) = bestObjs(i)
  	  }
  	  bestObj = currentObjective.value
  	  bound = bestObj+1
  	  tightenedOnce(currentObjectiveIdx) = true
  	}
	
  	override def reTighten() = {
  	  bound = bestObj+1
  	}  	
	
	override def relax() = {
	  bound = Int.MinValue
	  bestObj = Int.MinValue
	}
	
	override def isOptimum() = {
	  (0 until bounds.size).forall(i => bounds(i) > upperBounds(i))
	}

	override def propagate(): CPOutcome = {
		for(i <- 0 until objVars.size) {
		  if (objVars(i).updateMin(bounds(i)) == CPOutcome.Failure) {
			return CPOutcome.Failure
		  }
		}
		CPOutcome.Suspend;
	}
	
	override def restoreBest() = {
	  bound = bestObj
	}	

}